src/Mono.Android/Xamarin.Android.Net/ServerCertificateCustomValidator.cs

```diff
@@ -0,0 +1,190 @@
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Security;
+using System.Security.Cryptography.X509Certificates;
+
+using Javax.Net.Ssl;
+
+using JavaCertificateException = Java.Security.Cert.CertificateException;
+using JavaX509Certificate = Java.Security.Cert.X509Certificate;
+
+namespace Xamarin.Android.Net
+{
+	internal sealed class ServerCertificateCustomValidator
+	{
+		public IHostnameVerifier HostnameVerifier => AlwaysAcceptingHostnameVerifier.Instance;
+
+		public Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> Callback { get; set; }
+
+		public ServerCertificateCustomValidator (Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> callback)
+		{
+			Callback = callback;
+		}
+
+		public ITrustManager[] ReplaceX509TrustManager (ITrustManager[]? trustManagers, HttpRequestMessage requestMessage)
+		{
+			var originalX509TrustManager = FindX509TrustManager(trustManagers);
+			var trustManagerWithCallback = new TrustManager (originalX509TrustManager, requestMessage, Callback);
+			return ModifyTrustManagersArray (trustManagers, original: originalX509TrustManager, replacement: trustManagerWithCallback);
+		}
+
+		private sealed class TrustManager : Java.Lang.Object, IX509TrustManager
+		{
+			private readonly IX509TrustManager? _internalTrustManager;
+			private readonly HttpRequestMessage _request;
+			private readonly Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> _serverCertificateCustomValidationCallback;
+
+			public TrustManager (
+				IX509TrustManager? internalTrustManager,
+				HttpRequestMessage request,
+				Func<HttpRequestMessage, X509Certificate2?, X509Chain?, SslPolicyErrors, bool> serverCertificateCustomValidationCallback)
+			{
+				_request = request;
+				_internalTrustManager = internalTrustManager;
+				_serverCertificateCustomValidationCallback = serverCertificateCustomValidationCallback;
+			}
+
+			public void CheckServerTrusted (JavaX509Certificate[] javaChain, string authType)
+			{
+				var sslPolicyErrors = SslPolicyErrors.None;
+
+				try {
+					_internalTrustManager?.CheckServerTrusted (javaChain, authType);
+				} catch (JavaCertificateException) {
+					sslPolicyErrors |= SslPolicyErrors.RemoteCertificateChainErrors;
+				}
+
+				var certificates = Convert (javaChain);
+				X509Certificate2? certificate = null;
+
+				if (certificates.Length > 0) {
+					certificate = certificates [0];
+				} else {
+					sslPolicyErrors |= SslPolicyErrors.RemoteCertificateNotAvailable;
+				}
+
+				if (!VerifyHostname (javaChain)) {
+					sslPolicyErrors |= SslPolicyErrors.RemoteCertificateNameMismatch;
+				}
+
+				if (!_serverCertificateCustomValidationCallback (_request, certificate, CreateChain (certificates), sslPolicyErrors)) {
+					throw new JavaCertificateException ("The remote certificate was rejected by the provided RemoteCertificateValidationCallback.");
```

Comment:
@simonrozsival: Do you know the "context"/calling method stack of how this method is invoked?  Specifically, can this exception type "leak" from/through e.g. `HttpClient.Send()`?

We can't have Java exception types "leak through" .NET methods which document the exception types that can be thrown; see e.g. [the `$(AndroidBoundExceptionType)` docs](https://docs.microsoft.com/en-us/xamarin/android/deploy-test/building-apps/build-properties#androidboundexceptiontype) and 60363ef4.

We probably *don't* want to throw `Java.Security.Cert.CertificateException` here, *unless* Java code is calling this `CheckServerTrusted()` and needs an appropriate Java exception for handling purposes.

Reply:
This method is invoked from Java (`SSLContext`) and throwing this specific exception is the way to reject the server's certificate (see the method signature https://developer.android.com/reference/javax/net/ssl/X509TrustManager#checkServerTrusted(java.security.cert.X509Certificate[],%20java.lang.String)). We never call this method from C#.